pallets/domains/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+        LowerLevelDomainsNotAllowed,
+        /// This account is not allowed to update the domain metadata.
+        NotDomainOwner,
+        /// Outer value exceeds its length limit.
+        OuterValueOffLengthLimit,
+        /// A new outer value is the same as the old one.
+        OuterValueNotChanged,
+        /// Reservation period cannot be a zero value.
+        ZeroReservationPeriod,
+        /// Cannot store a domain for that long period of time.
+        TooBigReservationPeriod,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::weight(<T as Config>::WeightInfo::register_domain())]
+        pub fn register_domain(
+            origin: OriginFor<T>,
+            target: <<T as frame_system::pallet::Config>::Lookup as StaticLookup>::Source,
+            full_domain: DomainName<T>,
+            content: Content,
+            expires_in: <T as frame_system::pallet::Config>::BlockNumber,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+            let owner = <T as frame_system::pallet::Config>::Lookup::lookup(target)?;
+
+            ensure!(!expires_in.is_zero(), Error::<T>::ZeroReservationPeriod);
+            ensure!(
+                expires_in <= T::ReservationPeriodLimit::get(),
+                Error::<T>::TooBigReservationPeriod,
+            );
+
+            // Note that while upper and lower case letters are allowed in domain
+            // names, domain names are not case-sensitive. That is, two names with
+            // the same spelling but different case are to be treated as if identical.
+            let domain_lc = Self::lower_domain_then_bound(full_domain.clone());
+
+            ensure!(!Self::reserved_domain(&domain_lc), Error::<T>::DomainIsReserved);
+
+            ensure_content_is_valid(content.clone())?;
+
+            Self::ensure_valid_domain(&domain_lc)?;
+
+            ensure!(
+                Self::registered_domain(&domain_lc).is_none(),
+                Error::<T>::DomainAlreadyOwned,
+            );
+
+            let domains_per_account = Self::registered_domains_by_owner(&owner).len();
+            ensure!(
+                domains_per_account < T::MaxDomainsPerAccount::get() as usize,
+                Error::<T>::TooManyDomainsPerAccount,
+            );
+
+            let expires_at = expires_in.saturating_add(System::<T>::block_number());
+
+            let deposit = T::DomainDeposit::get();
+            let domain_meta = DomainMeta::new(
+                expires_at,
+                owner.clone(),
+                content,
+                deposit,
+            );
+
+            <T as Config>::Currency::reserve(&owner, deposit)?;
+
+            // TODO: withdraw balance
+
+            RegisteredDomains::<T>::insert(domain_lc.clone(), domain_meta);
+            RegisteredDomainsByOwner::<T>::mutate(
+                &owner, |domains| {
+                    domains.try_push(domain_lc.clone()).expect("qed; too many domains per account")
```

Comment:
We throw `TooManyDomainsPerAccount` before, because at this point we would need to unreserve funds and ensure we haven't written anything else into any other storage. We could add that error here as well, but it will be redundant.